staging/src/k8s.io/apimachinery/pkg/api/meta/help.go

```diff
@@ -114,6 +114,18 @@ func getItemsPtr(list runtime.Object) (interface{}, error) {
 // EachListItem invokes fn on each runtime.Object in the list. Any error immediately terminates
 // the loop.
 func EachListItem(obj runtime.Object, fn func(runtime.Object) error) error {
+	return eachListItem(obj, fn, false)
+}
+
+// EachListItemWithAlloc invokes fn on each runtime.Object in the list.
+// This differs from EachListItem by passing shallow-copies of struct items to `fn`
+// to avoid retaining references to memory belonging to the list.
+func EachListItemWithAlloc(obj runtime.Object, fn func(runtime.Object) error) error {
+	return eachListItem(obj, fn, true)
+}
+
+// allocNew: Whether shallow copy is required when the elements in Object.Items are struct
+func eachListItem(obj runtime.Object, fn func(runtime.Object) error, allocNew bool) error {
 	if unstructured, ok := obj.(runtime.Unstructured); ok {
 		return unstructured.EachListItem(fn)
```

Comment:
do we need the same fix in unstructured.EachListItem?

Reply:
it looks like UnstructuredList#EachListItem could very efficiently avoid referencing the list memory like this:

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go
index 5028f5fb57a..0196f3f4dfe 100644
--- a/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go
+++ b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_list.go
@@ -45,7 +45,7 @@ func (u *UnstructuredList) IsList() bool { return true }
 
 func (u *UnstructuredList) EachListItem(fn func(runtime.Object) error) error {
 	for i := range u.Items {
-		if err := fn(&u.Items[i]); err != nil {
+		if err := fn(&Unstructured{Object: u.Items[i].Object}); err != nil {
 			return err
 		}
 	}
```

that's actually what Unstructured#EachListItem already does

Reply:
Modifying UnstructuredList#EachListItem would break the constraints of the API. So adjust the whole method behavior by modifying the `fn` passed in. PTAL.